### ISSUE TITLE
Check session endpoint enabled, but SessionId is missing

### ISSUE THREAD

**Issue by parkinsona** [IdentityServer/IdentityServer3#2333]
We have had identityserver in production for about 6 months and just recently (past month and a half) it has started to return the error: "Check session endpoint enabled, but SessionId is missing".

This is only happening in our production environment. Which was moved to a different data centre around the same time the issue started happening. Having said that, we had an increase in usage of the site at the same time too. Both our UAT and PROD environments are load balanced and were moved at the same time.

We haven't been able to replicate it, but we know that it happens when the idsvr cookie is still active, but the idsvr.session cookie is dead. I'm not sure how it could get into this state though.I am yet to determine if its being caused by the "remember me" functionality or not.

This is the setting for cookie options in our IdentityServer host. Expire Time span is 1 hour.

``` c#
CookieOptions = new Thinktecture.IdentityServer.Core.Configuration.CookieOptions   
                        {
                            SlidingExpiration = true,                           
                            ExpireTimeSpan = Settings.Default.ExpireTimeSpan
                        }
```
Does this cookie setting cover both the idsvr & idsvr.session cookies that are created? Is it possible that the idsvr cookie is sliding but the idsvr.session cookie is not?

Is there a reason why identityserver doesn't redirect you to the login page when you get into this situation?

Thanks again for your help and support on this product.

**Comment by parkinsona** [comment]
I should add that i'm running version 1.6.0.0 of identity server.

**Comment by parkinsona** [comment]
I have further feedback from one of the users experiencing this issue....

They are using IE 11 and this is how they replicated it.

* Go to RP1 and hit login
* Login to identity server and select "Remember me"
* They are logged in and redirected to RP1
* They work in RP1 for a while, then close their browser
* Later in the day they open RP2, and click login
* They get redirected to identity server, but it will return an empty session id.

**Comment by brockallen** [comment]
Ok, will look into this.

**Comment by parkinsona** [comment]
I hope you have a break over christmas ;)

**Comment by brockallen** [comment]
Took a quick look -- it does look like we're issuing a persistent session cookie along with the persistent authentication cookie. It's working here (of course).

So what it sounds like is that your users keep one cookie but not the other. Is that correct?

I'd need some help from your side to help debug this some more.

**Comment by parkinsona** [comment]
On my machine, I have tested it and it in firefox is shows that it issues a persistent cookie for both session and authentication. On another machine using firefox, it is only showing as a session cookie. But we weren't able to replicate the issue using that.

From what I can see it looks like its keeping the authentication cookie but not the session cookie.

It wouldn't be related to this katana bug? 
https://katanaproject.codeplex.com/workitem/197

**Comment by brockallen** [comment]
> It wouldn't be related to this katana bug? 

Only if you're mixing System.Web cookie code in the same request with OWIN/Katana cookie code. Node of the code in IdSvr uses System.Web. Do you have other code in the IdSvr host using System.Web? A http module perhaps?

**Comment by parkinsona** [comment]
I'm not using System.web cookie code in my host.
I do use System.web and other httpmodules for things like logging, and my password reset view (which is used as a partial login)

**Comment by brockallen** [comment]
Yea, this might be hard to track down then where/what's causing it since neither you or I can repro it ourselves. For all we know, it's the user browser throwing away a cookie, or maybe a load balancer not passing a cookie thru... again, difficult without a repro.

**Comment by parkinsona** [comment]
Agreed, without the repro I've been struggling. I've only seen it happen once on the users machine, and clearing the cookies and re-logging fixed the issue.

Are there any known issues with load balancers not passing through cookies?

So if we can't track down the cause, should identity server bail at this point and redirect you to the login screen?

**Comment by brockallen** [comment]
I've heard of LBs dropping cookies when there are too many. But yea, I think that's rare or really just a fluke. 

If there's no session at token issuance time, then we can gen one. That might be what we should do, but doing this is somewhat like saying "somewhere else cookies aren't working for some reason, so let's cover it up by doing what should have been done elsewhere". 

I don't like that it's just not working right, so I'd like to know why before I put in code to mask other problems,

**Comment by brockallen** [comment]
But... come to think of it, I might like to move this to the authorization endpoint, since it is an aspect of the OIDC protocol. I'll have to think about it.

**Comment by parkinsona** [comment]
I agree that covering it up might not be the best option. 
I am running version 1.6, so its not something that may have been fixed in a newer version? From what I can see, the handling of the session cookie hasn't changed.

The only other thing to mention, is that our sys admin has got the app pool recycling at 3am every night. I wouldn't think this would cause any dramas because our user is only using the site during business hours.

**Comment by brockallen** [comment]
no, apppool is unrelated. session id is a random value, not related to any keys or anything

**Comment by parkinsona** [comment]
That's what I thought, just wanted to check.

Here's my startup code, if it helps....


``` c#
 public void Configuration(IAppBuilder app)
        {
            LogProvider.SetCurrentLogProvider(new DiagnosticsTraceLogProvider());
            //LogProvider.SetCurrentLogProvider(new TraceSourceLogProvider());

            // uncomment to enable HSTS headers for the host
            // see: https://developer.mozilla.org/en-US/docs/Web/Security/HTTP_strict_transport_security
            //app.UseHsts();

            app.Map("/core", coreApp =>
                {

                    LoginPageLink resetLink = new LoginPageLink() { Href = Settings.Default.PasswordResetLink, Text="Forgot Password?" };
                    
                    var options = new IdentityServerOptions
                    {
                        
                        IssuerUri = "https://mysite.com.au",                       
                        SiteName = "MySite IdentityServer",
                        
                        PublicOrigin = Settings.Default.PublicHostName,
                        SigningCertificate = Cert.Load(),
                        Factory = Factory.Configure("MembershipReboot"), // This is the name of the connection string
                        CorsPolicy = CorsPolicy.AllowAll,

                         
                        RequireSsl = true,
                        AuthenticationOptions = new AuthenticationOptions
                        {
                           // IdentityProviders = ConfigureIdentityProviders,
                            LoginPageLinks = new LoginPageLink[] { resetLink },
                        CookieOptions = new Thinktecture.IdentityServer.Core.Configuration.CookieOptions   
                        {
                            SlidingExpiration = true,                           
                            ExpireTimeSpan = Settings.Default.ExpireTimeSpan
                        }
                        }

                    };

                    coreApp.UseIdentityServer(options);
                });

            // only for showing the getting started index page
            app.UseStaticFiles();

            //Use Log4Net to log events
            XmlConfigurator.Configure();            
            LogProvider.SetCurrentLogProvider(new Log4NetLogProvider());
            

        }
```

**Comment by parkinsona** [comment]
Hi Brock, did you have any more thoughts on this issue?

**Comment by brockallen** [comment]
No, not really. @leastprivilege and I are meeting next week, so we can discuss then.

**Comment by parkinsona** [comment]
Hey Brock,

I have some more info regarding this issue.
We had another user have the issue today (11/1).
This time it was for another relying party. After speaking with 2 users of this application, 1 used "Remember Me" and the other didn't. The one who didn't use "Remember Me" was able to log in today, the one who did couldn't log in.

Both were using the same RP, IE 11 and were accessing the site at the same time. 
Both could log in on (8/11)
This fits with our previous case, where the only user having the issue was one who chose the "Remember me" option.

Unfortunately, they cleared their cookies before I could investigate the issue. So that's all I have.

**Comment by brockallen** [comment]
I tried to repro it again here locally with IE11 and I can't get the behavior you described (when I check "remember my login").

**Comment by brockallen** [comment]
Also, just to clarify; you said:

> We have had identityserver in production for about 6 months and just recently (past month and a half) it has started to return the error: "Check session endpoint enabled, but SessionId is missing".

But it's not preventing users from logging in, right? It's just not issuing a `session_state` value in the response, correct?

**Comment by brockallen** [comment]
Also, are you doing any custom login screen where you're taking over issuing the login cookie?

**Comment by parkinsona** [comment]
Hi Brock,

It is preventing users from logging in. It's as though they  browser loses the session_state cookie after a certain period of time. To allow them to log in again, they've been clearing their cookies and starting authentication again.

I have custom styles for the login screen and I have a custom partial sign in page that will prompt users to reset their passwords when they have expired.

**Comment by brockallen** [comment]
Where or how exactly does it prevent them from logging in? Does IDSvr throw an exception?

**Comment by parkinsona** [comment]
When the RP redirects to idsvr, idsvr throws an exception.

Here's what's happening in the log prior: 

``` 

2016-01-25 15:37:47,884 [56] INFO  Thinktecture.IdentityServer.Core.Endpoints.AuthorizeEndpointController - Start authorize request


2016-01-25 15:37:47,884 [56] INFO  Thinktecture.IdentityServer.Core.Validation.AuthorizeRequestValidator - Start authorize request protocol validation


2016-01-25 15:37:47,900 [56] ERROR Thinktecture.IdentityServer.Core.Validation.AuthorizeRequestValidator - Check session endpoint enabled, but SessionId is missing
 {
  "ClientId": "mysite_implicitclient",
  "ClientName": "mysite",
  "RedirectUri": "https://mysite.com.au/",
  "AllowedRedirectUris": [
    "https://mysite.com.au/"
  ],
  "SubjectId": "64fbc3c2-86bb-41b8-b141-f97540c7aa18",
  "ResponseType": "id_token",
  "ResponseMode": "form_post",
  "Flow": "Implicit",
  "RequestedScopes": "openid email profile myscope",
  "State": "OpenIdConnect.AuthenticationProperties=g5XIMum2EI1Q3LvrgvwVWmCVqTM8QQzhQAlYseGUsnoFS9Ogs7cj5NrXnlmjpQEgOMJy7f4Re5cugVM2_aavm_zgof5-78vo6Uc42OvgRGORJz3Rqx8ATM8-_BvkRdJdZdr5sD_lDZi4v3Zz2DwPvgAqPtEc5uDCyckfkP5RU1Vf3moAiCQyWhQMnLgSD3YlcTWX31vsIuvFYDSSSHyzZ_PLZKRyQFOaXWd619dQU14",
  "Nonce": "635892934679230492.MzhmNGQ2Y2UtNDJiNS00M2E4LWFiMWMtODE4NjdiNmU5Y2EwNThiMjZmODgtNmQ5ZS00MTZlLTg4ODMtMjQ2ZWM5NjkxMTli",
  "Raw": {
    "client_id": "mysite_implicitclient",
    "redirect_uri": "https://mysite.com.au/",
    "response_mode": "form_post",
    "response_type": "id_token",
    "scope": "openid email profile myscope",
    "state": "OpenIdConnect.AuthenticationProperties=g5XIMum2EI1Q3LvrgvwVWmCVqTM8QQzhQAlYseGUsnoFS9Ogs7cj5NrXnlmjpQEgOMJy7f4Re5cugVM2_aavm_zgof5-78vo6Uc42OvgRGORJz3Rqx8ATM8-_BvkRdJdZdr5sD_lDZi4v3Zz2DwPvgAqPtEc5uDCyckfkP5RU1Vf3moAiCQyWhQMnLgSD3YlcTWX31vsIuvFYDSSSHyzZ_PLZKRyQFOaXWd619dQU14",
    "nonce": "635892934679230492.MzhmNGQ2Y2UtNDJiNS00M2E4LWFiMWMtODE4NjdiNmU5Y2EwNThiMjZmODgtNmQ5ZS00MTZlLTg4ODMtMjQ2ZWM5NjkxMTli"
  }
}

```

I'll add some screenshots in another message.

**Comment by brockallen** [comment]
What version of IdSvr? It looks like 1.x something?

**Comment by parkinsona** [comment]
I'm running 1.6.0.0 in PROD.
We have version 2.3 in DEV, which I'm waiting on some internal approvals to deploy through to TEST and UAT to see if we can replicate in the latest version.

**Comment by brockallen** [comment]
> When the RP redirects to idsvr, idsvr throws an exception.

That's a log error, not a runtime exception. I'm still unclear how users are prevented from logging in... 

But anyway, perhaps 2.3 is your best best on this -- is it possible to test if your 2.3 dev is still having this issue?

**Comment by parkinsona** [comment]
Sorry, I had that wrong. It's not idsvr with the exception, its the RP.  
The RP checks for the sessionState claim, and throws an exception when it doesn't find it. The RP is using the sessionState claim for single sign out of the RPs.

**Comment by brockallen** [comment]
So again, I'm not sure what we should do with this. Is there any way you can reproduce this with our samples?

**Comment by parkinsona** [comment]
Unfortunately I've not been able to replicate it with samples. 
I know we had discussed adding something to handle the scenario when no session cookie was present, but you had to think more about that.

If worst comes to worst, I can turn off the "remember me" option.In the meantime I'm happy to wait this out and see if it continues to happen in PROD after I update to the latest release of idsvr. This might be a few weeks away with my current release schedules.

**Comment by brockallen** [comment]
Ideally you would be able to repro with our samples. If not then perhaps you can slowly add your custom code to our samples to determine the trigger point so we understand what's causing it. 

-Brock

**Comment by parkinsona** [comment]
I can't even repro with my own code. It seems to only be happening for a few users, and it only happens about once a month.Unfortunately those users aren't on site, so often they have cleared cookies before I get the chance to see what's happening in their browser. I've instructed them to let me know when it happens.

**Comment by brockallen** [comment]
Ok, got it. Well, keep us posted if you learn more. 

-Brock

**Comment by parkinsona** [comment]
Absolutely :+1: 

Thanks again for your help looking at this.

**Comment by parkinsona** [comment]
We had this issue this morning with an external customer. Unfortunately we couldn't troubleshoot with them because they just wanted to login again.

However, I can confirm that it is definitely the remember me option in IE.

We noticed a few extra things this time round. The user had changed their password just last night, so the cookie hadn't expired. Your previous comments mentioned using System.Web cookie, which I thought maybe the partial login to change password might be causing it, although  this doesn't use cookies. I was also able to disprove this theory because the last time we had this issue the user hadn't changed their password for quite some time.

I thought I'd have another look for cookies and I did find this in the web.config. I must have dismissed it first time because it was in config.

```
  <system.web>
     ....
    <httpCookies httpOnlyCookies="true" requireSSL="true" />
    ....
  </system.web>
```

Interestingly.....the idsvr.session cookie is the only cookie being issued that is not HttpOnly. I'm not sure if that could be the cause?
![check_session_cookies](https://cloud.githubusercontent.com/assets/8508456/12834082/9e904f4c-cbf0-11e5-8baa-c31f1e782f5f.png)

**Comment by parkinsona** [comment]
As an FYI, I tried removing the httpCookies setting from my web.config and it produced the same results.

**Comment by parkinsona** [comment]
One of our internal users had the issue yesterday and I was able to hop on their machine and have a look around. Attached is the xml from IE's F12, which can be loaded into fiddler.

From what I can tell, the idsvr.session cookie just isn't getting passed through from the RP to idsvr.
[NetworkData_20160208.zip](https://github.com/IdentityServer/IdentityServer3/files/122319/NetworkData_20160208.zip)

**Comment by brockallen** [comment]
Ok, took a look. The log actually starts a bit too late -- it shows after the login has taken place, so we don't see the response from the login page where the session id should have been issued.

The one thing cookie related that looks odd to me is this cookie:

```
<cookie>
 <name>_ga</name>
 <value>GA1.3.1517871657.1451868525</value>
</cookie>
```

What this tells me is that there's something in your web server or network that's adding more cookies. Perhaps it's somehow involved with the session cookie not making it back to the browser?

**Comment by parkinsona** [comment]
I thought the same about the log being too late. But that's when the issue occurs. When they initially login to idsvr they're all good and can work and there is a session cookie. I know that the session cookie exists initially because the RPs implement checkSession iframe for single signout. Without the idsvr.session cookie they would be logged out automatically after 30 seconds of being on the RP. At least that's my understanding of how that works.

The issue happens a few days after they have been using the site for a few days. In the morning they'll open their browser, go to the RP and then it will give you the log I sent through. To me, it seems like the browser might be ditching the cookie.

I thought the _ga cookie was odd as well. It supposedly for Google Analytics. However, to my knowledge, none of our sites use Google Analytics. I know for certain that idsvr doesn't use GA. I will need to confirm this in the other RPs.

**Comment by brockallen** [comment]
Come to think of it, the other cookie i don't see is the antixsrf cookie.

**Comment by brockallen** [comment]
Maybe it's related to the fact there are dots in those cookies....?

**Comment by parkinsona** [comment]
That's a good point about the antixsrf.
I just looked at my current session in firefox and I had idsvr and idsvr.session cookies, but not the antixsrf cookie. I always thought it was short lived and only used for the initial login, which is why it was gone all the time. 

However, if its supposed to be there. Perhaps, there's something else at play.

**Comment by brockallen** [comment]
If they just had logged in, then I'd expect to see the antixsrf cookie.

**Comment by parkinsona** [comment]
As the users affected are selecting "Remember Me", they could have logged in a day ago, or 3 weeks ago.

**Comment by parkinsona** [comment]
Hey Brock,

I can confirm that none of our RPs are using Google analytics, so I'm not sure why that's being included. Perhaps something in the browser is keeping that ga cookie around.

I think you have a point about both the antixsrf and session cookies both having a "." has some merit. I checked on another pc we've got setup to catch this error, and even after several days all 3 cookies are there (idsvr, idsvr.session, idsvr.xsrf).

**Comment by brockallen** [comment]
The other thing I was wondering is if these users have any browser plugins that might be affecting things.... just another thought.

**Comment by parkinsona** [comment]
I've requested a list of plugins used by the user. I'm waiting on a response. Will keep you posted.

**Comment by brockallen** [comment]
Sorry this is taking so long to resolve, but i'm 99% sure there's something environmental.

**Comment by parkinsona** [comment]
It's all good. It's not your fault its taking so long. The main person who has the issue only works a couple of days a week, so it takes a while to get responses on my end.

I'm pretty sure its environmental as well.

**Comment by parkinsona** [comment]
Hey Brock,

I got the feedback about the add-ons in IE.
They were:

- Shockwave Flash Object 
- Send to OneNote
- OneNote Linked Notes

There were also some disabled add-ons:
- Skype for Business Browser Hel....
- Skype for Business Click to Call
- Symantec Vulnerability Protection

hmmmmmm I wonder... When we were working on our bulk upload, one of the other developers had an issue with identity server and skype. Let me chase it up with him and get back to you. Could be a while because he only works 1 day a week :S

**Comment by parkinsona** [comment]
I have a feeling it might have been the skype app running on their pc though.

**Comment by parkinsona** [comment]
I've had one of my internal users notify me of this issue this morning.
It was the same as the previous user. Log and Addons are attached.

However, I've noticed a pattern for my internal users. It seems to start happening on a Monday.
Although for client users it has been on other days. I know in our environment there are a lot of maintenance tasks that run over the weekends. I'll ask around and see if there's anything that runs that could interfere with this or the load balancer.

We're scheduled to release version 2.3 of idsvr to Production in the next couple of days, so I think we will keep an eye on how it progresses after the deployment.
[rk_identity_ie_issue_log.zip](https://github.com/IdentityServer/IdentityServer3/files/160629/rk_identity_ie_issue_log.zip)

**Comment by parkinsona** [comment]
I finally managed to release v2.3 to our PROD environment this week.
I will keep a lookout to see if this problem is resolved with the newer version of idsvr

**Comment by brockallen** [comment]
How's it going?

**Comment by parkinsona** [comment]
So far so good.
From what I've noticed, they normally happening in the first week of the month, so I'll see how we go next week.

I did get some check session errors about a day after I deployed v2.3. I was kind of writing those off because those cookies would have been created from the old version.

**Comment by parkinsona** [comment]
Unfortunately, in the past few days the errors have begun to resurface.
I found #2717 and that sounds like a probable cause for the issue I'm facing.

**Comment by brockallen** [comment]
So then I'd suggest trying to add custom MW to add the necessary headers. It'll be much faster for you to test it with custom MW than to wait for me.

**Comment by parkinsona** [comment]
I'll look into it.

**Comment by parkinsona** [comment]
I can confirm that this error is still present. I'm not sure the MW will fix the problem because I'm hosting on the same domain. I could make the change, but I still can't replicate it locally, so I wouldn't know if it was fixed for about a month or so by the time existing cookies expire.

**Comment by parkinsona** [comment]
Another update, I have managed to get this error to occur locally. I had not previously been able to do that. This rules out the possibility of the cookie being dropped from the load balancer.

**Comment by brockallen** [comment]
I spoke to another customer last week and they were finding they were losing cookies due to to many cookies on the same host. Just another data point for consideration.

**Comment by parkinsona** [comment]
Thanks for the update on this brock. I have been on leave for several months due to family health issues.
I was speaking to some colleagues last week when I returned to work, and it sounds like they might have been able to replicate this issue. 

I'll keep you posted.

**Comment by Stitchup** [comment]
Hi brock, I'm a colleague of parkinsona and have been looking into this issue further as we have a client requesting persistent login. I've been able to replicate this issue and it seems to relate back to parkinsona's early suspicion that the idsvr cookie is sliding but the idsvr.session cookie not. It relates to those situations where persistent cookies are issued (RememberMe selected or (external login with IsPersistent option = true)) and the SlidingExpiration option = true.
An example cookie options configuration might be -
CookieOptions = new IdentityServer3.Core.Configuration.CookieOptions   {
    SlidingExpiration = true,
    IsPersistent = true,
    AllowRememberMe = true,
    RememberMeDuration = TimeSpan.FromMinutes(10),
    ExpireTimeSpan = TimeSpan.FromMinutes(5)
}


There is potentially a window where the user idsvr.session cookie has expired but the idsvr cookie has not. After login both cookies will have the same initial expiry time. The idsvr cookie will slide if the user is active beyond a threshold (50%?). Should the user then attempt to access a resource after both cookies have expired the user is identified as logged out, as expected. If however the user attempts to access a resource after the idsvr.session cookie has expired (still at the initial expiry time) but before the idsvr cookie expires (as this time may have slid if user was active) then an attempt is made to access the sessionId value from the idsvr.session cookie which for all intensive purposes does not exist (as it has expired).

**Comment by thebodaciousd** [comment]
I am also seeing this issue with a couple of our users.  Based on the couple of issues I have sen with this error, I am trying to get more information from the 4 users (out of 1500) who are experiencing this problem   to track down the root cause.  Anyone have anything definitive they have found I can check for first?

**Comment by thebodaciousd** [comment]
After looking through our scenarios, I can confirm @Stitchup and @parkinsona findings.  If sliding expiration is used, the authentication cookie will get extended but the session cookie does not.  We first noticed because our client applications try to pull out the sid claim and use the same value in the client applications as the session identifier so we can trace user activity throughout applications.  Even though the error is logged in idsvr, the authentication actually goes through to the client, minus the sid claim.   Unless someone comes up with a better idea, we are just implementing a short lifetime on the client application cookies (60 minutes), but then using a persistent cookie in Identity Server (30 days) to get the effect of a sliding expiration.

**Comment by ArturDorochowicz** [comment]
I believe I've been experiencing the same issue. We noticed on the client side, which in our case is Katana OpenIdConnectAuthenticationMiddleware, that `OpenIdConnectMessage.SessionState` property is null for some users, despite the fact that all endpoints in IdentityServer are enabled (as they are by default).

We are on IdentityServer3 2.5.1 currently.

Interestingly, we also have `CookieOptions.SlidingExpiration` set to true for IdentityServer.

**Comment by jskahjem** [comment]
We also have this issue, and I can confirm the findings of @Stitchup, @parkinsona and @thebodaciousd. Our app is a singlepage app that uses IdServer as a as an STS. Our app is also using persistent, sliding cookies. In our case we have calls to refresh the token periodically, which keeps the idsrv cookie alive beyond the initial expiry date. After looking at the cookies in IdServer and looking at the error log I see that once the idsrv.session cookie disappears (since it's not sliding), we get these errors in the log (from the next call to refresh token and for all calls after this). 

We're using version 2.5.0.

**Comment by tbrettallen** [comment]
This issue is also prevalent for me. The strange part is, that is that we aren't experiencing any complaints and there is no issue during normal usage visible to end users from what we're seeing.

We're basically just ignoring these log issues for now since they happen constantly.

We are using a custom sign-in page. Here is the code:

```
    public class SignInController : Controller
    {
        private static ILogger Log = LogManager.GetCurrentClassLogger();

        public ActionResult Index(string id)
        {
            try
            {
                if (string.IsNullOrEmpty(id))
                    return new HttpStatusCodeResult(HttpStatusCode.BadRequest, "Id is required.");

                var env = Request.GetOwinContext().Environment;
                var signInMessage = env.GetSignInMessage(id);

                if (signInMessage == null)
                    throw new HttpResponseException(HttpStatusCode.BadRequest);

                return View(signInMessage);
            }
            catch (Exception ex)
            {
                Log.Error(ex, LogTemplates.ErrorOccurredDuringImplicitSignIn);
                throw;
            }
        }

        [System.Web.Mvc.HttpPost]
        public ActionResult Index(string id, string username, string password, string tenant, string rememberMe)
        {
            try
            {
                var env = Request.GetOwinContext().Environment;
                var signInMessage = env.GetSignInMessage(id);
                signInMessage.Tenant = tenant;

                var result = UserService.Authenticate(username, password, signInMessage.IdP, tenant);

                if (result == null || result.IsError)
                {
                    Response.TrySkipIisCustomErrors = true;
                    Response.StatusCode = (int)HttpStatusCode.BadRequest;

                    return result == null ? PartialView("_LoginErrorNullAuthenticateResult", false) : PartialView("_LoginError", result);
                }

                env.IssueLoginCookie(new IdentityServer3.Core.Models.AuthenticatedLogin
                {
                    Subject = result.User.GetSubjectId(),
                    Claims = result.User.Claims,
                    IdentityProvider = signInMessage.IdP,
                    Name = result.User.GetName(),
                    AuthenticationMethod = result.User.GetAuthenticationMethod(),
                    PersistentLogin = Convert.ToBoolean(rememberMe ?? false.ToString()),
                    PersistentLoginExpiration = DateTimeOffset.UtcNow.AddDays(14)
                });

                var msg = env.GetSignInMessage(id);
                var returnUrl = msg.ReturnUrl;

                env.RemovePartialLoginCookie();

                return Json(returnUrl);
            }
            catch (Exception ex)
            {
                Log.Error(ex, LogTemplates.ErrorOccurredDuringImplicitAuthentication);
                throw;
            }
        }
    }
```

![logs](https://cloud.githubusercontent.com/assets/3762313/25008055/1eb50a8c-2028-11e7-90f2-be2fa6e65fd5.png)

**Comment by brockallen** [comment]
Basically the issue is that tracking one logical thing in two cookies, and so when they're not in sync this happens. Since we're not planning on making future enhancements here, I can only suggest that you filter this error message from your logging.
